dali/test/python/test_operator_numba_func.py

```diff
@@ -0,0 +1,150 @@
+    batch_size = len(shapes)
+    pipe = numba_func_pipe(batch_size=batch_size, num_threads=1, device_id=0, shapes=shapes, dtype=dtype, fn_ptr=fn_ptr, setup_fn=setup_fn)
+    pipe.build()
+    outs = pipe.run()
+    for _ in range(3):
+        for i in range(batch_size):
+            out_arr = np.array(outs[0][i])
+            assert np.array_equal(out_arr, expected_out[i])
+
+def test_numba_func():
+    # shape, dtype, func address, expected_out
+    args = [
+        ([(10, 10, 10)], np.uint8, set_all_values_to_255.address, None, [np.full((10, 10, 10), 255, dtype=np.uint8)]),
+        ([(10, 10, 10)], np.float32, set_all_values_to_float.address, None, [np.full((10, 10, 10), 0.5, dtype=np.float32)]),
+        ([(10, 20, 30), (20, 10, 30)], np.int64, fun3.address, setup_fn1.address, [np.full((20, 30, 10), 42, dtype=np.int32), np.full((30, 20, 10), 42, dtype=np.int32)]),
+    ]
+
+    for shape, dtype, fn_ptr, setup_fn, expected_out in args:
+        yield _testimpl_numba_func, shape, dtype, fn_ptr, setup_fn, expected_out
+
+@pipeline_def
+def numba_func_image_pipe(fn_ptr=None, setup_fn=None):
+    files, _ = dali.fn.readers.caffe(path=lmdb_folder)
+    images_in = dali.fn.decoders.image(files, device="cpu")
+    images_out = dali.fn.experimental.numba_func(images_in, fn_ptr=fn_ptr, setup_fn=setup_fn)
+    return images_in, images_out
+
+def _testimpl_numba_func_image(fn_ptr, setup_fn, transform):
+    pipe = numba_func_image_pipe(batch_size=8, num_threads=1, device_id=0, fn_ptr=fn_ptr, setup_fn=setup_fn)
+    pipe.build()
+    for _ in range(3):
+        images_in, images_out = pipe.run()
+        for i in range(len(images_in)):
+            image_in_transformed = transform(images_in.at(i))
+            assert np.array_equal(image_in_transformed, images_out.at(i))
+
+@cfunc(fun1_sig, nopython=True)
+def reverse_col(out_ptr, out_shape_ptr, in_ptr, in_shape_ptr, ndim):
+    out_shape = carray(out_shape_ptr, ndim)
+    in_shape = carray(in_shape_ptr, ndim)
+    in_arr = carray(in_ptr, (in_shape[0], in_shape[1], in_shape[2]))
+    out_arr = carray(out_ptr, (out_shape[0], out_shape[1], out_shape[2]))
+    out_arr[:] = 255 - in_arr[:]
+
+@cfunc(fun1_sig, nopython=False)
+def rot_image(out_ptr, out_shape_ptr, in_ptr, in_shape_ptr, ndim):
+    out_shape = carray(out_shape_ptr, ndim)
+    in_shape = carray(in_shape_ptr, ndim)
+    in_arr = carray(in_ptr, (in_shape[0], in_shape[1], in_shape[2]))
+    out_arr = carray(out_ptr, (out_shape[0], out_shape[1], out_shape[2]))
+    for i in range(out_shape[0]):
+        for j in range(out_shape[1]):
+            out_arr[i][j] = in_arr[j][out_shape[0] - i - 1]
+
+@cfunc(setup1_sig, nopython=True)
+def rot_image_setup(out_shape_ptr, out1_ndim, out_dtype, in_shape_ptr, in1_ndim, in_dtype, num_samples, num_outputs, num_inputs):
+    in_arr = carray(in_shape_ptr, num_samples * out1_ndim)
+    out_arr = carray(out_shape_ptr, num_samples * in1_ndim)
```

Comment:
I wonder if we can have some helper functions to deal with this repeating pattern (similar in the case of the function itself).

Reply:
I think in this case it's not needed. It will be as complicated as invoking carray function.